File: app/src/main/java/wangdaye/com/geometricweather/background/interfaces/CMWeatherProviderService.java
package wangdaye.com.geometricweather.background.interfaces;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import cyanogenmod.providers.WeatherContract;
import cyanogenmod.weather.RequestInfo;
import cyanogenmod.weather.WeatherInfo;
import cyanogenmod.weatherservice.ServiceRequest;
import cyanogenmod.weatherservice.ServiceRequestResult;
import cyanogenmod.weatherservice.WeatherProviderService;
import dagger.hilt.android.AndroidEntryPoint;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.location.LocationHelper;
import wangdaye.com.geometricweather.settings.SettingsManager;
import wangdaye.com.geometricweather.weather.WeatherHelper;

/**
 * CM weather provider service.
 * */

@AndroidEntryPoint
public class CMWeatherProviderService extends WeatherProviderService
        implements WeatherHelper.OnRequestWeatherListener {

    @Nullable
    private ServiceRequest mRequest;

    @Inject LocationHelper mLocationHelper;
    @Inject WeatherHelper mWeatherHelper;

    private final LocationHelper.OnRequestLocationListener locationListener
            = new LocationHelper.OnRequestLocationListener() {

        @Override
        public void requestLocationSuccess(Location requestLocation) {
            if (mRequest != null) {
                mWeatherHelper.requestWeather(
                        CMWeatherProviderService.this,
                        requestLocation,
                        CMWeatherProviderService.this);
            }
        }

        @Override
        public void requestLocationFailed(Location requestLocation) {
            if (mRequest != null) {
                mRequest.fail();
            }
        }
    };

    private final WeatherHelper.OnRequestLocationListener weatherLocationListener
            = new WeatherHelper.OnRequestLocationListener() {
        @Override
        public void requestLocationSuccess(String query, List<Location> locationList) {
            if (mRequest != null) {
                if (locationList != null && locationList.size() > 0) {
                    mWeatherHelper.requestWeather(
                            CMWeatherProviderService.this,
                            locationList.get(0),
                            CMWeatherProviderService.this);
                } else {
                    requestLocationFailed(query);
                }
            }
        }

        @Override
        public void requestLocationFailed(String query) {
            if (mRequest != null) {
                mRequest.fail();
            }
        }
    };

    @Override
    public void onCreate() {
        super.onCreate();
        mRequest = null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        cancelRequest();
    }

    @Override
    protected void onRequestSubmitted(ServiceRequest serviceRequest) {
        cancelRequest();
        mRequest = serviceRequest;

        RequestInfo info = serviceRequest.getRequestInfo();
        switch (info.getRequestType()) {
            case RequestInfo.TYPE_WEATHER_BY_WEATHER_LOCATION_REQ:
                requestWeather(info.getWeatherLocation().getCity());
                break;

            case RequestInfo.TYPE_WEATHER_BY_GEO_LOCATION_REQ:
                requestLocation();
                break;

            case RequestInfo.TYPE_LOOKUP_CITY_NAME_REQ:
                requestWeather(info.getCityName());
                break;

            default:
                serviceRequest.fail();
                break;
        }
    }

    @Override
    protected void onRequestCancelled(ServiceRequest serviceRequest) {
        cancelRequest();
    }

    // control.

    private void requestLocation() {
        mLocationHelper.requestLocation(
                this,
                Location.buildLocal(),
                true,
                locationListener
        );
    }

    private void requestWeather(String cityName) {
        if (!TextUtils.isEmpty(cityName)) {
            List<WeatherSource> list = new ArrayList<>();
            list.add(SettingsManager.getInstance(this).getWeatherSource());
            mWeatherHelper.requestLocation(this, cityName, list, weatherLocationListener);
        } else if (mRequest != null) {
            mRequest.fail();
        }
    }

    private void cancelRequest() {
        mRequest = null;
        mLocationHelper.cancel();
        mWeatherHelper.cancel();
    }

    // interface.

    // on request weather listener.

    @Override
    public void requestWeatherSuccess(@NonNull Location requestLocation) {
        try {
            Weather weather = requestLocation.getWeather();
            if (mRequest != null && weather != null) {
                List<WeatherInfo.DayForecast> forecastList = new ArrayList<>();
                for (int i = 0; i < weather.getDailyForecast().size(); i ++) {
                    forecastList.add(
                            new WeatherInfo.DayForecast.Builder(
                                    WeatherConditionConvertHelper.getConditionCode(
                                            weather.getDailyForecast().get(i).day().getWeatherCode(),
                                            true
                                    ))
                                    .setHigh(weather.getDailyForecast().get(i).day().getTemperature().getTemperature())
                                    .setHigh(weather.getDailyForecast().get(i).night().getTemperature().getTemperature())
                                    .build()
                    );
                }
                WeatherInfo.Builder builder = new WeatherInfo.Builder(
                        requestLocation.getCityName(getApplicationContext()),
                        weather.getCurrent().getTemperature().getTemperature(),
                        WeatherContract.WeatherColumns.TempUnit.CELSIUS
                ).setWeatherCondition(
                                WeatherConditionConvertHelper.getConditionCode(
                                        weather.getCurrent().getWeatherCode(),
                                        requestLocation.isDaylight()
                                )
                ).setTodaysHigh(
                        weather.getDailyForecast().get(0).day().getTemperature().getTemperature()
                ).setTodaysLow(
                        weather.getDailyForecast().get(0).night().getTemperature().getTemperature()
                ).setTimestamp(
                        weather.getBase().getTimeStamp()
                );
                if (weather.getCurrent().getRelativeHumidity() != null) {
                    builder.setHumidity(weather.getCurrent().getRelativeHumidity());
                }
                if (weather.getCurrent().getWind().getSpeed() != null) {
                    builder.setWind(
                            weather.getCurrent().getWind().getSpeed(),
                            weather.getCurrent().getWind().getDegree().getDegree(),
                            WeatherContract.WeatherColumns.WindSpeedUnit.KPH
                    ).setForecast(forecastList);
                }

                mRequest.complete(new ServiceRequestResult.Builder(builder.build()).build());
            }
        } catch (Exception ignore) {
            requestWeatherFailed(requestLocation, false);
        }
    }

    @Override
    public void requestWeatherFailed(@NonNull Location requestLocation, @NonNull Boolean apiLimitReached) {
        if (mRequest != null) {
            mRequest.fail();
        }
    }
}

class WeatherConditionConvertHelper {
    static int getConditionCode(WeatherCode code, boolean dayTime) {
        switch (code) {
            case CLEAR:
                if (dayTime) {
                    return WeatherContract.WeatherColumns.WeatherCode.SUNNY;
                } else {
                    return WeatherContract.WeatherColumns.WeatherCode.CLEAR_NIGHT;
                }

            case PARTLY_CLOUDY:
                if (dayTime) {
                    return WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_DAY;
                } else {
                    return WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_NIGHT;
                }

            case CLOUDY:
                return WeatherContract.WeatherColumns.WeatherCode.CLOUDY;

            case RAIN:
                return WeatherContract.WeatherColumns.WeatherCode.SHOWERS;

            case SNOW:
                return WeatherContract.WeatherColumns.WeatherCode.SNOW;

            case WIND:
                return WeatherContract.WeatherColumns.WeatherCode.WINDY;

            case FOG:
                return WeatherContract.WeatherColumns.WeatherCode.FOGGY;

            case HAZE:
                return WeatherContract.WeatherColumns.WeatherCode.HAZE;

            case SLEET:
                return WeatherContract.WeatherColumns.WeatherCode.SLEET;

            case HAIL:
                return WeatherContract.WeatherColumns.WeatherCode.HAIL;

            case THUNDER:
                return WeatherContract.WeatherColumns.WeatherCode.THUNDERSTORMS;

            case THUNDERSTORM:
                return WeatherContract.WeatherColumns.WeatherCode.THUNDERSHOWER;
        }
        return WeatherContract.WeatherColumns.WeatherCode.NOT_AVAILABLE;
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/background/interfaces/CMWeatherProviderService.kt
package wangdaye.com.geometricweather.background.interfaces

import android.text.TextUtils
import cyanogenmod.providers.WeatherContract
import cyanogenmod.weather.RequestInfo
import cyanogenmod.weather.WeatherInfo
import cyanogenmod.weather.WeatherInfo.DayForecast
import cyanogenmod.weatherservice.ServiceRequest
import cyanogenmod.weatherservice.ServiceRequestResult
import cyanogenmod.weatherservice.WeatherProviderService
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.*
import wangdaye.com.geometricweather.GeometricWeather
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode
import wangdaye.com.geometricweather.location.LocationHelper
import wangdaye.com.geometricweather.settings.SettingsManager
import wangdaye.com.geometricweather.weather.WeatherHelper
import java.util.*
import javax.inject.Inject

/**
 * CM weather provider service.
 */
@AndroidEntryPoint
class CMWeatherProviderService : WeatherProviderService() {

    private var request: ServiceRequest? = null

    @Inject lateinit var locationHelper: LocationHelper
    @Inject lateinit var weatherHelper: WeatherHelper

    private var job: Job? = null

    private object WeatherConditionConvertHelper {
        fun getConditionCode(code: WeatherCode?, dayTime: Boolean): Int {
            when (code) {
                WeatherCode.CLEAR -> return if (dayTime) {
                    WeatherContract.WeatherColumns.WeatherCode.SUNNY
                } else {
                    WeatherContract.WeatherColumns.WeatherCode.CLEAR_NIGHT
                }
                WeatherCode.PARTLY_CLOUDY -> return if (dayTime) {
                    WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_DAY
                } else {
                    WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_NIGHT
                }
                WeatherCode.CLOUDY -> return WeatherContract.WeatherColumns.WeatherCode.CLOUDY
                WeatherCode.RAIN -> return WeatherContract.WeatherColumns.WeatherCode.SHOWERS
                WeatherCode.SNOW -> return WeatherContract.WeatherColumns.WeatherCode.SNOW
                WeatherCode.WIND -> return WeatherContract.WeatherColumns.WeatherCode.WINDY
                WeatherCode.FOG -> return WeatherContract.WeatherColumns.WeatherCode.FOGGY
                WeatherCode.HAZE -> return WeatherContract.WeatherColumns.WeatherCode.HAZE
                WeatherCode.SLEET -> return WeatherContract.WeatherColumns.WeatherCode.SLEET
                WeatherCode.HAIL -> return WeatherContract.WeatherColumns.WeatherCode.HAIL
                WeatherCode.THUNDER -> return WeatherContract.WeatherColumns.WeatherCode.THUNDERSTORMS
                WeatherCode.THUNDERSTORM -> return WeatherContract.WeatherColumns.WeatherCode.THUNDERSHOWER
            }
            return WeatherContract.WeatherColumns.WeatherCode.NOT_AVAILABLE
        }
    }

    override fun onCreate() {
        super.onCreate()
        request = null
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelRequest()
    }

    override fun onRequestSubmitted(serviceRequest: ServiceRequest) {
        cancelRequest()

        request = serviceRequest

        val info = serviceRequest.requestInfo
        when (info.requestType) {
            RequestInfo.TYPE_WEATHER_BY_WEATHER_LOCATION_REQ -> requestUpdate(info.weatherLocation.city)
            RequestInfo.TYPE_WEATHER_BY_GEO_LOCATION_REQ -> requestUpdate()
            RequestInfo.TYPE_LOOKUP_CITY_NAME_REQ -> requestUpdate(info.cityName)
            else -> serviceRequest.fail()
        }
    }

    override fun onRequestCancelled(serviceRequest: ServiceRequest) {
        cancelRequest()
    }

    // control.

    private fun requestUpdate(cityName: String? = null) = GeometricWeather.instance.applicationScope.launch {
        val context = this@CMWeatherProviderService

        val location = withContext(Dispatchers.IO) {

            if (TextUtils.isEmpty(cityName)) {
                val response = locationHelper.getLocation(
                        context, Location.buildLocal(), true)
                response.result
            } else {
                val response = weatherHelper.getLocation(
                        context,
                        cityName!!,
                        listOf(SettingsManager.getInstance(context).getWeatherSource())
                )
                response.result?.let { it[0] }
            }
        }

        location?.let {
            it.weather = withContext(Dispatchers.IO) {
                weatherHelper.getWeather(context, it).result
            }
            completeRequest(it)
            return@launch
        }

        request?.fail()
    }

    private fun completeRequest(requestLocation: Location) {
        try {
            val weather = requestLocation.weather
            if (request != null && weather != null) {
                val forecastList: MutableList<DayForecast> = ArrayList()
                for (i in weather.dailyForecast.indices) {
                    forecastList.add(
                            DayForecast.Builder(
                                    WeatherConditionConvertHelper.getConditionCode(
                                            weather.dailyForecast[i].day().weatherCode,
                                            true
                                    )
                            ).setHigh(weather.dailyForecast[i].day().temperature.temperature.toDouble())
                                    .setHigh(weather.dailyForecast[i].night().temperature.temperature.toDouble())
                                    .build()
                    )
                }
                val builder = WeatherInfo.Builder(
                        requestLocation.getCityName(applicationContext),
                        weather.current.temperature.temperature.toDouble(),
                        WeatherContract.WeatherColumns.TempUnit.CELSIUS
                ).setWeatherCondition(
                        WeatherConditionConvertHelper.getConditionCode(
                                weather.current.weatherCode,
                                requestLocation.isDaylight
                        )
                ).setTodaysHigh(weather.dailyForecast[0].day().temperature.temperature.toDouble())
                        .setTodaysLow(weather.dailyForecast[0].night().temperature.temperature.toDouble())
                        .setTimestamp(weather.base.timeStamp)
                if (weather.current.relativeHumidity != null) {
                    builder.setHumidity(weather.current.relativeHumidity!!.toDouble())
                }
                if (weather.current.wind.speed != null) {
                    builder.setWind(
                            weather.current.wind.speed!!.toDouble(),
                            weather.current.wind.degree.degree.toDouble(),
                            WeatherContract.WeatherColumns.WindSpeedUnit.KPH
                    ).setForecast(forecastList)
                }
                request?.complete(ServiceRequestResult.Builder(builder.build()).build())
            }
        } catch (ignore: Exception) {
            request?.fail()
        }
    }

    private fun cancelRequest() {
        request = null

        job?.cancel()
        job = null
    }
}